Provides support for installing on vanilla kubernetes

phantomjinx opened the pull request:
- Upgrades the deprecated `api/batch/v1beta1` to `api/batch/v1`;
- Formats test results using `gotestfmt`; 
- Provides `capabilities` API for discovering the type of cluster (OpenShift or Kubernetes) based on the existence of installed APIs, also providing the Kubernetes version and/or the OpenShift release version;
- Tidies up function calls by wrapping parameter variables into structs;
- Provides self-signed certificate creation and installation for the `hawtio-online-tls-serving` secret required for k8;
- `cr` and `uninstall` rules for Makefiles;
- Creation of `ingress` for Kubernetes and `route` for OpenShift;

---
**Should be merged / rebased after**
~1. https://github.com/hawtio/hawtio-operator/pull/101~
~2. https://github.com/hawtio/hawtio-operator/pull/102~

phantomjinx committed:
Introduces capabilities module for determining cluster particulars #100

* APICapabilities determines cluster version and existence of certain
  APIs, eg. Routes. From this determination, it concludes whether the
  cluster is OpenShift4 compatible

* Allows a clear determination as to whether this is an OpenShift cluster
  or vanilla Kubernetes. This can be used to make choices on what to
  install, eg. Route or Ingress

* Have the cluster capabilities evaluated on initialising of the operator

phantomjinx committed:
Modifies the certificate function API to clarify its purpose #100

* The function generateCertificateSecret can be used for both OpenShift
  CA signing and self-signing certificates. Just a case of determining
  what is used as both the caCertificate and caPrivateKey (self-signing
  uses the generated certificate and private key)

* Rather than using the same function with the caSecret as nil, makes it
  obvious which to use by providing front functions that describe which
  sort of certificate secret will be created

	install-operator.sh

phantomjinx committed:
fix: Adds uninstall rule to Makefile #100

* Removes all of setup, operator and app resources

phantomjinx committed:
Adds rule to Makefile to install only the CR #100

phantomjinx committed:
Start to modify the watches to evaluate what is present in the cluster #100

* Only evaluate the Route API if it is present in the cluster.

phantomjinx committed:
Start to wrap variables into struct properties #100

* hawtio_controller.go
 * Rather than adding to the extensive list of function parameters, roll
   them up into a single struct that can be populated then passed around
   instead

* deployment.go
 * Pass in the capabilities.ApiSpec that wraps up both the IsOpenShift and
   version properties

phantomjinx committed:
feat: Adds notional support for extracting the url into the hawtio.Status #100

* While the extraction of the route url is quite straightforward, the
  use of ingress path rules and no hosts makes the return of a single
  string quite difficult. This may well need work in the future to
  improve the value returned

tadayosi reviewed (changes requested):
Can you please split the mere dependency updates from the substantial changes?

1731 files changes make it impossible to review it on GitHub as the browser becomes frozen...

tadayosi commented:
You can also format test code separately from the issue.

phantomjinx commented:
Checks will fail until https://github.com/hawtio/hawtio-operator/pull/101 has been merged.

phantomjinx committed:
Implements support for installation on kubernetes #100

* Chooses between route (openshift-only) and ingress API types

* Installs self-signed certificate for kubernetes installs for the
  hawtio-online-serving secret

* Only reconcile an OAuth client if on OpenShift for now

* Ensures the HAWTIO_ONLINE_AUTH envvar is always populated to specify the
  authentication type rather than relying on the fallback in the actual
  hawtio-online image config.sh script

* Differentiate between openshift and k8 code by putting in different files
  and packages if possible
